File: idi/training/python/idi_iann/env.py
"""Synthetic environment harness for Q-table training."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Tuple
import random

from .config import QuantizerConfig, RewardWeights


@dataclass(frozen=True)
class EnvObservation:
    """Quantized snapshot returned by the simulator."""

    price_bucket: int
    volume_bucket: int
    trend_bucket: int
    scarcity_bucket: int
    mood_bucket: int

    def as_state(self) -> Tuple[int, int, int, int, int]:
        return (
            self.price_bucket,
            self.volume_bucket,
            self.trend_bucket,
            self.scarcity_bucket,
            self.mood_bucket,
        )


class EmoteChannel:
    """Tracks high-level mood for emoji/text generation."""

    def __init__(self, max_bucket: int):
        self._max = max_bucket
        self._value = 0

    def update(self, reward_signal: float) -> int:
        if reward_signal > 0.5:
            self._value = min(self._max - 1, self._value + 1)
        elif reward_signal < -0.2:
            self._value = max(0, self._value - 1)
        return self._value


@dataclass
class SyntheticMarketEnv:
    """Very small stochastic market emulator tailored for lookup tables."""

    ACTIONS = ("hold", "buy", "sell")

    quantizer: QuantizerConfig
    rewards: RewardWeights
    seed: int = 0
    _random: random.Random = field(init=False, repr=False)
    _emote: EmoteChannel = field(init=False, repr=False)
    _tick: int = field(default=0, init=False, repr=False)
    _position: int = field(default=0, init=False, repr=False)  # -1, 0, 1

    def __post_init__(self) -> None:
        self._random = random.Random(self.seed)
        self._emote = EmoteChannel(self.quantizer.mood_buckets)

    def reset(self) -> EnvObservation:
        self._tick = 0
        self._position = 0
        self._emote = EmoteChannel(self.quantizer.mood_buckets)
        return self._observe()

    def step(self, action: str) -> Tuple[EnvObservation, float]:
        if action not in self.ACTIONS:
            raise ValueError(f"Unknown action {action}")
        self._apply_action(action)
        reward = self._compute_reward()
        self._tick += 1
        mood_bucket = self._emote.update(reward)
        obs = self._observe(overrides={"mood_bucket": mood_bucket})
        reward += self.rewards.communication_clarity * mood_bucket
        return obs, reward

    def _observe(self, overrides: dict | None = None) -> EnvObservation:
        price = self._random.randint(0, self.quantizer.price_buckets - 1)
        volume = self._random.randint(0, self.quantizer.volume_buckets - 1)
        trend = self._random.randint(0, self.quantizer.trend_buckets - 1)
        scarcity = self._random.randint(0, self.quantizer.scarcity_buckets - 1)
        mood = 0
        if overrides:
            mood = overrides.get("mood_bucket", mood)
        return EnvObservation(price, volume, trend, scarcity, mood)

    def _apply_action(self, action: str) -> None:
        delta = {"buy": 1, "sell": -1}.get(action, 0)
        self._position = max(-1, min(1, self._position + delta))

    def _compute_reward(self) -> float:
        price_drift = self._random.uniform(-1.0, 1.0)
        scarcity_change = self._random.uniform(-0.5, 0.5)
        pnl_reward = self._position * price_drift * self.rewards.pnl
        scarcity_bonus = scarcity_change * self.rewards.scarcity_alignment
        ethics_bonus = self.rewards.ethics_bonus if self._position >= 0 else -self.rewards.ethics_bonus
        return pnl_reward + scarcity_bonus + ethics_bonus

File: idi/training/rust/idi_iann/src/env.rs
use rand::{Rng, SeedableRng};

use crate::action::Action;
use crate::config::{QuantizerConfig, RewardWeights};
use crate::traits::{Environment, Observation};

#[derive(Clone, Copy)]
pub struct EnvObservation {
    pub price: u8,
    pub volume: u8,
    pub trend: u8,
    pub scarcity: u8,
    pub mood: u8,
}

impl Observation for EnvObservation {
    fn as_state(&self) -> (u8, u8, u8, u8, u8) {
        (
            self.price,
            self.volume,
            self.trend,
            self.scarcity,
            self.mood,
        )
    }
}

#[derive(Clone)]
pub struct SyntheticMarketEnv {
    quant: QuantizerConfig,
    rewards: RewardWeights,
    rng: rand::rngs::StdRng,
    position: i8,
}

impl SyntheticMarketEnv {
    pub const ACTIONS: [Action; 3] = [Action::Hold, Action::Buy, Action::Sell];

    pub fn new(quant: QuantizerConfig, rewards: RewardWeights, seed: u64) -> Self {
        Self {
            quant,
            rewards,
            rng: SeedableRng::seed_from_u64(seed),
            position: 0,
        }
    }

    fn observe(&mut self) -> EnvObservation {
        EnvObservation {
            price: self.rng.gen_range(0..self.quant.price_buckets),
            volume: self.rng.gen_range(0..self.quant.volume_buckets),
            trend: self.rng.gen_range(0..self.quant.trend_buckets),
            scarcity: self.rng.gen_range(0..self.quant.scarcity_buckets),
            mood: self.rng.gen_range(0..self.quant.mood_buckets),
        }
    }

    fn update_position(&mut self, action: Action) {
        match action {
            Action::Buy => self.position = (self.position + 1).clamp(-1, 1),
            Action::Sell => self.position = (self.position - 1).clamp(-1, 1),
            Action::Hold => {}
        }
    }

    fn calculate_reward(&self, price_drift: f32, scarcity_drift: f32) -> f32 {
        let pnl = price_drift * self.position as f32 * self.rewards.pnl;
        let scarcity_bonus = scarcity_drift * self.rewards.scarcity_alignment;
        let ethics_bonus = if self.position >= 0 {
            self.rewards.ethics_bonus
        } else {
            -self.rewards.ethics_bonus
        };
        pnl + scarcity_bonus + ethics_bonus
    }
}

impl Environment for SyntheticMarketEnv {
    type Obs = EnvObservation;

    /// Reset environment to initial state.
    ///
    /// # Returns
    /// Initial observation with position = 0
    fn reset(&mut self) -> Self::Obs {
        self.position = 0;
        self.observe()
    }

    /// Execute action and return next observation and reward.
    ///
    /// # Arguments
    /// * `action` - Trading action (Hold/Buy/Sell)
    ///
    /// # Returns
    /// Tuple of (next_observation, reward)
    ///
    /// # Invariants
    /// * Position is clamped to [-1, 1]
    /// * Reward components: PnL + scarcity + ethics
    fn step(&mut self, action: Action) -> (Self::Obs, f32) {
        self.update_position(action);
        let price_drift: f32 = self.rng.gen_range(-1.0..1.0);
        let scarcity_drift: f32 = self.rng.gen_range(-0.5..0.5);
        let reward = self.calculate_reward(price_drift, scarcity_drift);
        (self.observe(), reward)
    }
}
